Leftover resources on provisioning failure

**Describe the bug**
Currently if the packer-provisioned machine fails to reach running state (for example because of capacity issues) it won't be cleaned up by the plugin. 
This unfortunately results in leaking resources that we need to clean up manually, which is not ideal.

**To Reproduce**
Provision a machine that will fail the starting -> running transition eventually (ex. time out on clout-init > 10 minutes or provision in an AZ that has no capacity)

This will result in the following:
```
11:23:49  ==> ibmcloud-vpc.testing-farm: *************************************************************
11:23:49  ==> ibmcloud-vpc.testing-farm: * Initializing IBM Cloud Packer Plugin - VPC Infrastructure *
11:23:49  ==> ibmcloud-vpc.testing-farm: *************************************************************
11:23:49  ==> ibmcloud-vpc.testing-farm:
11:23:49  ==> ibmcloud-vpc.testing-farm: Creating VPC service...
11:23:49  ==> ibmcloud-vpc.testing-farm: VPC service creation successful!
11:23:49  ==> ibmcloud-vpc.testing-farm: Retrieving Subnet 0767-0458a498-e138-4e3a-b466-b74a9ac1cf38 information...
11:23:49  ==> ibmcloud-vpc.testing-farm: Subnet Information successfully retrieved ...
11:23:49  ==> ibmcloud-vpc.testing-farm: VPC ID: r014-37e76de5-4ab9-46c1-8161-405ec6eb53db
11:23:49  ==> ibmcloud-vpc.testing-farm: Zone: us-east-2
11:23:49  ==> ibmcloud-vpc.testing-farm: Fetching Base Image ID...
11:23:49  ==> ibmcloud-vpc.testing-farm: Base Image ID fetched: r014-90c06985-569b-4eb1-93eb-7a4f7c6a47dd
11:23:49  ==> ibmcloud-vpc.testing-farm: Creating SSH Public and Private Key Pair...
11:23:49  ==> ibmcloud-vpc.testing-farm: ed25519 public and private SSH key pair successfully created.
11:23:49  ==> ibmcloud-vpc.testing-farm: Creating a new SSH key for VPC...
11:23:49  ==> ibmcloud-vpc.testing-farm: SSH Key for VPC successfully created!
11:23:49  ==> ibmcloud-vpc.testing-farm: SSH Key for VPC's Name: packer-vpc-ssh-key-1750757023898444522
11:23:49  ==> ibmcloud-vpc.testing-farm: SSH Key for VPC's ID: r014-9e595e00-d553-4b8a-9a56-6103eaed5783
11:23:49  ==> ibmcloud-vpc.testing-farm: Creating Instance...
11:23:50  ==> ibmcloud-vpc.testing-farm: Instance successfully created!
11:23:50  ==> ibmcloud-vpc.testing-farm: Instance's Name: packer-vpc-vsi-1750757023898444522
11:23:50  ==> ibmcloud-vpc.testing-farm: Instance's ID: 0767_ec65dfe6-e9a0-4d53-9c84-aec229aba7c4
11:23:50  ==> ibmcloud-vpc.testing-farm: Waiting for the instance to become ACTIVE...
11:23:50  ==> ibmcloud-vpc.testing-farm: .
11:24:01  ==> ibmcloud-vpc.testing-farm: .
11:24:11  ==> ibmcloud-vpc.testing-farm: .
11:24:22  ==> ibmcloud-vpc.testing-farm: .
11:24:32  ==> ibmcloud-vpc.testing-farm: .
11:24:43  ==> ibmcloud-vpc.testing-farm: Waiting time: 1 minutes
11:24:53  ==> ibmcloud-vpc.testing-farm: .
11:25:04  ==> ibmcloud-vpc.testing-farm: .
11:25:15  ==> ibmcloud-vpc.testing-farm: .
11:25:24  ==> ibmcloud-vpc.testing-farm: .
11:25:34  ==> ibmcloud-vpc.testing-farm: .
11:25:45  ==> ibmcloud-vpc.testing-farm: Waiting time: 2 minutes
11:25:56  ==> ibmcloud-vpc.testing-farm: .
11:26:05  ==> ibmcloud-vpc.testing-farm: .
11:26:15  ==> ibmcloud-vpc.testing-farm: .
11:26:26  ==> ibmcloud-vpc.testing-farm: .
11:26:40  ==> ibmcloud-vpc.testing-farm: .
11:26:45  ==> ibmcloud-vpc.testing-farm: Waiting time: 3 minutes
11:26:56  ==> ibmcloud-vpc.testing-farm: .
11:27:06  ==> ibmcloud-vpc.testing-farm: .
11:27:16  ==> ibmcloud-vpc.testing-farm: .
11:27:27  ==> ibmcloud-vpc.testing-farm: .
11:27:37  ==> ibmcloud-vpc.testing-farm: .
11:27:48  ==> ibmcloud-vpc.testing-farm: Waiting time: 4 minutes
11:27:58  ==> ibmcloud-vpc.testing-farm: .
11:28:08  ==> ibmcloud-vpc.testing-farm: .
11:28:18  ==> ibmcloud-vpc.testing-farm: .
11:28:33  ==> ibmcloud-vpc.testing-farm: .
11:28:40  ==> ibmcloud-vpc.testing-farm: .
11:28:48  ==> ibmcloud-vpc.testing-farm: Waiting time: 5 minutes
11:28:59  ==> ibmcloud-vpc.testing-farm: .
11:28:59  ==> ibmcloud-vpc.testing-farm: [ERROR] Error step waiting for instance to become ACTIVE: [ERROR] Instance return with failed status. Status Reason - failed: Can't start instance because resource capacity is unavailable.
11:28:59  ==> ibmcloud-vpc.testing-farm: Deleting SSH key for VPC packer-vpc-ssh-key-1750757023898444522 ...
11:29:32  ==> ibmcloud-vpc.testing-farm: [ERROR] Error deleting SSH key for VPC packer-vpc-ssh-key-1750757023898444522. Please delete it manually: SSH-key still in use by 1 instance(s). First instance ID: 0767_ec65dfe6-e9a0-4d53-9c84-aec229aba7c4
11:29:32  ==> ibmcloud-vpc.testing-farm: Deleting Public and Private SSH Key Pair...
11:29:32  ==> ibmcloud-vpc.testing-farm: Public and Private SSH Key Pair successfully deleted.
11:29:32  ==> ibmcloud-vpc.testing-farm:
11:29:32  ==> ibmcloud-vpc.testing-farm: ********************************************************************
11:29:32  ==> ibmcloud-vpc.testing-farm: * Thank you for using IBM Cloud Packer Plugin - VPC Infrastructure *
11:29:32  ==> ibmcloud-vpc.testing-farm: ********************************************************************
11:29:32  Build 'ibmcloud-vpc.testing-farm' errored after 5 minutes 45 seconds: unexpected EOF
```

**Expected behavior**
All packer-provisioned resources are cleaned up.

--- a/builder/ibmcloud/vpc/step_create_instance.go
+++ b/builder/ibmcloud/vpc/step_create_instance.go
@@ -370,34 +370,36 @@ func (step *stepCreateInstance) Cleanup(state multistep.StateBag) {
 
 	// Delete Floating IP if it was created (VSI Interface was set as public)
 	if config.VSIInterface == "public" {
-		floatingIP := state.Get("floating_ip").(string)
-		ui.Say(fmt.Sprintf("Releasing the Floating IP: %s ...", floatingIP))
+		if state.Get("floating_ip") != nil {
+			floatingIP := state.Get("floating_ip").(string)
+			ui.Say(fmt.Sprintf("Releasing the Floating IP: %s ...", floatingIP))
 
-		floatingIPID := state.Get("floating_ip_id").(string)
+			floatingIPID := state.Get("floating_ip_id").(string)
 
-		options := vpcService.NewGetFloatingIPOptions(floatingIPID)
-		floatingIPresponse, _, err := vpcService.GetFloatingIP(options)
-		if err != nil {
-			err := fmt.Errorf("[ERROR] Error getting the Floating IP: %s", err)
-			state.Put("error", err)
-			ui.Error(err.Error())
-			// log.Fatalf(err.Error())
-			return
-		}
-		status := floatingIPresponse.Status
-		if *status == "available" {
-			options := vpcService.NewDeleteFloatingIPOptions(floatingIPID)
-			result, err := vpcService.DeleteFloatingIP(options)
-
+			options := vpcService.NewGetFloatingIPOptions(floatingIPID)
+			floatingIPresponse, _, err := vpcService.GetFloatingIP(options)
 			if err != nil {
-				err := fmt.Errorf("[ERROR] Error releasing the Floating IP. Please release it manually: %s", err)
+				err := fmt.Errorf("[ERROR] Error getting the Floating IP: %s", err)
 				state.Put("error", err)
 				ui.Error(err.Error())
 				// log.Fatalf(err.Error())
 				return
 			}
-			if result.StatusCode == 204 {
-				ui.Say("The Floating IP was successfully released!")
+			status := floatingIPresponse.Status
+			if *status == "available" {
+				options := vpcService.NewDeleteFloatingIPOptions(floatingIPID)
+				result, err := vpcService.DeleteFloatingIP(options)
+
+				if err != nil {
+					err := fmt.Errorf("[ERROR] Error releasing the Floating IP. Please release it manually: %s", err)
+					state.Put("error", err)
+					ui.Error(err.Error())
+					// log.Fatalf(err.Error())
+					return
+				}
+				if result.StatusCode == 204 {
+					ui.Say("The Floating IP was successfully released!")
+				}
 			}
 		}
 	}
@@ -440,24 +442,25 @@ func (step *stepCreateInstance) Cleanup(state multistep.StateBag) {
 	}
 
 	// Deleting Security Group's rule
+	if state.Get("security_group_rule_id") != nil {
+		ruleID := state.Get("security_group_rule_id").(string)
+		ui.Say(fmt.Sprintf("Deleting Security Group's rule %s ...", ruleID))
+		sgRuleOptions := &vpcv1.DeleteSecurityGroupRuleOptions{}
+		sgRuleOptions.SetSecurityGroupID(state.Get("security_group_id").(string))
+		sgRuleOptions.SetID(ruleID)
+		sgRuleResponse, sgRuleErr := vpcService.DeleteSecurityGroupRule(sgRuleOptions)
 
-	ruleID := state.Get("security_group_rule_id").(string)
-	ui.Say(fmt.Sprintf("Deleting Security Group's rule %s ...", ruleID))
-	sgRuleOptions := &vpcv1.DeleteSecurityGroupRuleOptions{}
-	sgRuleOptions.SetSecurityGroupID(state.Get("security_group_id").(string))
-	sgRuleOptions.SetID(ruleID)
-	sgRuleResponse, sgRuleErr := vpcService.DeleteSecurityGroupRule(sgRuleOptions)
+		if sgRuleErr != nil {
+			sgRuleErr := fmt.Errorf("[ERROR] Error deleting Security Group's rule %s. Please delete it manually: %s", ruleID, sgRuleErr)
+			state.Put("error", sgRuleErr)
+			ui.Error(err.Error())
+			// log.Fatalf(err.Error())
+			return
+		}
 
-	if sgRuleErr != nil {
-		sgRuleErr := fmt.Errorf("[ERROR] Error deleting Security Group's rule %s. Please delete it manually: %s", ruleID, sgRuleErr)
-		state.Put("error", sgRuleErr)
-		ui.Error(err.Error())
-		// log.Fatalf(err.Error())
-		return
-	}
-
-	if sgRuleResponse.StatusCode == 204 {
-		ui.Say("The Security Group's rule was successfully deleted!")
+		if sgRuleResponse.StatusCode == 204 {
+			ui.Say("The Security Group's rule was successfully deleted!")
+		}
 	}
 
 	// Wait a couple of seconds before attempting to delete the security group.
@@ -465,22 +468,24 @@ func (step *stepCreateInstance) Cleanup(state multistep.StateBag) {
 
 	// Deleting Security Group
 	if config.SecurityGroupID == "" {
-		securityGroupName := state.Get("security_group_name").(string)
-		ui.Say(fmt.Sprintf("Deleting Security Group %s ...", securityGroupName))
-		securityGroupID := state.Get("security_group_id").(string)
-		sgOptions := &vpcv1.DeleteSecurityGroupOptions{}
-		sgOptions.SetID(securityGroupID)
-		sgResponse, err := vpcService.DeleteSecurityGroup(sgOptions)
-		if err != nil {
-			err := fmt.Errorf("[ERROR] Error deleting Security Group %s. Please delete it manually: %s", securityGroupName, err)
-			state.Put("error", err)
-			ui.Error(err.Error())
-			// log.Fatalf(err.Error())
-			return
-		}
+		if state.Get("security_group_name") != nil {
+			securityGroupName := state.Get("security_group_name").(string)
+			ui.Say(fmt.Sprintf("Deleting Security Group %s ...", securityGroupName))
+			securityGroupID := state.Get("security_group_id").(string)
+			sgOptions := &vpcv1.DeleteSecurityGroupOptions{}
+			sgOptions.SetID(securityGroupID)
+			sgResponse, err := vpcService.DeleteSecurityGroup(sgOptions)
+			if err != nil {
+				err := fmt.Errorf("[ERROR] Error deleting Security Group %s. Please delete it manually: %s", securityGroupName, err)
+				state.Put("error", err)
+				ui.Error(err.Error())
+				// log.Fatalf(err.Error())
+				return
+			}
 
-		if sgResponse.StatusCode == 204 {
-			ui.Say("The Security Group was successfully deleted!")
+			if sgResponse.StatusCode == 204 {
+				ui.Say("The Security Group was successfully deleted!")
+			}
 		}
 	}
 

--- a/builder/ibmcloud/vpc/step_waitfor_instance.go
+++ b/builder/ibmcloud/vpc/step_waitfor_instance.go
@@ -3,6 +3,7 @@ package vpc
 import (
 	"context"
 	"fmt"
+	"time"
 
 	"github.com/IBM/vpc-go-sdk/vpcv1"
 	"github.com/hashicorp/packer-plugin-sdk/multistep"
@@ -35,4 +36,127 @@ func (s *stepWaitforInstance) Run(_ context.Context, state multistep.StateBag) m
 	return multistep.ActionContinue
 }
 
-func (client *stepWaitforInstance) Cleanup(state multistep.StateBag) {}
+func (client *stepWaitforInstance) Cleanup(state multistep.StateBag) {
+	config := state.Get("config").(Config)
+	ui := state.Get("ui").(packer.Ui)
+	var vpcService *vpcv1.VpcV1
+	if state.Get("vpcService") != nil {
+		vpcService = state.Get("vpcService").(*vpcv1.VpcV1)
+	}
+
+	// Delete Floating IP if it was created (VSI Interface was set as public)
+	if config.VSIInterface == "public" {
+		floatingIP := state.Get("floating_ip").(string)
+		ui.Say(fmt.Sprintf("Releasing the Floating IP: %s ...", floatingIP))
+
+		floatingIPID := state.Get("floating_ip_id").(string)
+
+		options := vpcService.NewGetFloatingIPOptions(floatingIPID)
+		floatingIPresponse, _, err := vpcService.GetFloatingIP(options)
+		if err != nil {
+			err := fmt.Errorf("[ERROR] Error getting the Floating IP: %s", err)
+			state.Put("error", err)
+			ui.Error(err.Error())
+			// log.Fatalf(err.Error())
+			return
+		}
+		status := floatingIPresponse.Status
+		if *status == "available" {
+			options := vpcService.NewDeleteFloatingIPOptions(floatingIPID)
+			result, err := vpcService.DeleteFloatingIP(options)
+
+			if err != nil {
+				err := fmt.Errorf("[ERROR] Error releasing the Floating IP. Please release it manually: %s", err)
+				state.Put("error", err)
+				ui.Error(err.Error())
+				// log.Fatalf(err.Error())
+				return
+			}
+			if result.StatusCode == 204 {
+				ui.Say("The Floating IP was successfully released!")
+			}
+		}
+	}
+
+	// Wait a couple of seconds before attempting to delete the instance.
+	time.Sleep(2 * time.Second)
+	instanceData := state.Get("instance_data").(*vpcv1.Instance)
+	instanceID := *instanceData.ID
+	ui.Say(fmt.Sprintf("Deleting Instance ID: %s ...", instanceID))
+
+	options := &vpcv1.DeleteInstanceOptions{}
+	options.SetID(instanceID)
+	_, err := vpcService.DeleteInstance(options)
+
+	if err != nil {
+		err := fmt.Errorf("[ERROR] Error deleting the instance. Please delete it manually: %s", err)
+		state.Put("error", err)
+		ui.Error(err.Error())
+		// log.Fatalf(err.Error())
+		return
+	}
+	instanceDeleted := false
+	for !instanceDeleted {
+		options := &vpcv1.GetInstanceOptions{}
+		options.SetID(instanceID)
+		instance, response, err := vpcService.GetInstance(options)
+		if err != nil {
+			if response != nil && response.StatusCode == 404 {
+				ui.Say("Instance deleted Succesfully")
+				instanceDeleted = true
+				break
+			}
+			err := fmt.Errorf("[ERROR] Error getting the instance to check delete status. %s", err)
+			state.Put("error", err)
+			ui.Error(err.Error())
+		} else if instance != nil {
+			ui.Say(fmt.Sprintf("Instance status :-  %s", *instance.Status))
+		}
+		time.Sleep(10 * time.Second)
+	}
+
+	// Deleting Security Group's rule
+
+	ruleID := state.Get("security_group_rule_id").(string)
+	ui.Say(fmt.Sprintf("Deleting Security Group's rule %s ...", ruleID))
+	sgRuleOptions := &vpcv1.DeleteSecurityGroupRuleOptions{}
+	sgRuleOptions.SetSecurityGroupID(state.Get("security_group_id").(string))
+	sgRuleOptions.SetID(ruleID)
+	sgRuleResponse, sgRuleErr := vpcService.DeleteSecurityGroupRule(sgRuleOptions)
+
+	if sgRuleErr != nil {
+		sgRuleErr := fmt.Errorf("[ERROR] Error deleting Security Group's rule %s. Please delete it manually: %s", ruleID, sgRuleErr)
+		state.Put("error", sgRuleErr)
+		ui.Error(err.Error())
+		// log.Fatalf(err.Error())
+		return
+	}
+
+	if sgRuleResponse.StatusCode == 204 {
+		ui.Say("The Security Group's rule was successfully deleted!")
+	}
+
+	// Wait a couple of seconds before attempting to delete the security group.
+	time.Sleep(10 * time.Second)
+
+	// Deleting Security Group
+	if config.SecurityGroupID == "" {
+		securityGroupName := state.Get("security_group_name").(string)
+		ui.Say(fmt.Sprintf("Deleting Security Group %s ...", securityGroupName))
+		securityGroupID := state.Get("security_group_id").(string)
+		sgOptions := &vpcv1.DeleteSecurityGroupOptions{}
+		sgOptions.SetID(securityGroupID)
+		sgResponse, err := vpcService.DeleteSecurityGroup(sgOptions)
+		if err != nil {
+			err := fmt.Errorf("[ERROR] Error deleting Security Group %s. Please delete it manually: %s", securityGroupName, err)
+			state.Put("error", err)
+			ui.Error(err.Error())
+			// log.Fatalf(err.Error())
+			return
+		}
+
+		if sgResponse.StatusCode == 204 {
+			ui.Say("The Security Group was successfully deleted!")
+		}
+	}
+}

